src/sentry/workflow_engine/migrations/0012_migrate_metric_alerts.py

```diff
@@ -0,0 +1,206 @@
+# Generated by Django 5.1.1 on 2024-11-08 23:16
+
+from enum import Enum, IntEnum
+
+from django.apps.registry import Apps
+from django.db import migrations, models
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+
+from sentry.incidents.grouptype import MetricAlertFire
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBarApprox
+
+
+class PriorityLevel(IntEnum):
+    LOW = 25
+    MEDIUM = 50
+    HIGH = 75
+
+
+class DetectorPriorityLevel(IntEnum):
+    OK = 0
+    LOW = PriorityLevel.LOW
+    MEDIUM = PriorityLevel.MEDIUM
+    HIGH = PriorityLevel.HIGH
+
+
+class IncidentStatus(Enum):
+    OPEN = 1
+    CLOSED = 2
+    WARNING = 10
+    CRITICAL = 20
+
+
+class ActionType(models.TextChoices):
+    Notification = "SendNotificationAction"
+    TriggerWorkflow = "TriggerWorkflowAction"
+
+
+class DataSourceType(models.IntegerChoices):
+    SNUBA_QUERY_SUBSCRIPTION = 1
+    SNUBA_QUERY = 2
+
+
+class DataConditionGroupType(models.TextChoices):
+    ANY = "any"
+    ALL = "all"
+    NONE = "none"
+
+
+class AlertRuleStatus(Enum):
+    PENDING = 0
+    SNAPSHOT = 4
+    DISABLED = 5
+    NOT_ENOUGH_DATA = 6
+
+
+def migrate_metric_alerts(apps: Apps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    AlertRule = apps.get_model("sentry", "AlertRule")
+    AlertRuleProjects = apps.get_model("sentry", "AlertRuleProjects")
+    AlertRuleTrigger = apps.get_model("sentry", "AlertRuleTrigger")
+    AlertRuleTriggerAction = apps.get_model("sentry", "AlertRuleTriggerAction")
+    # AlertRuleActivity = apps.get_model("sentry", "AlertRuleActivity")
+    QuerySubscription = apps.get_model("sentry", "QuerySubscription")
+    Incident = apps.get_model("sentry", "Incident")
+
+    DataSource = apps.get_model("sentry", "DataSource")
+    DataConditionGroup = apps.get_model("sentry", "DataConditionGroup")
+    Workflow = apps.get_model("sentry", "Workflow")
+    Detector = apps.get_model("sentry", "Detector")
+    DetectorState = apps.get_model("sentry", "DetectorState")
+    DataCondition = apps.get_model("sentry", "DataCondition")
+    Action = apps.get_model("sentry", "Action")
+    DataConditionGroupAction = apps.get_model("sentry", "DataConditionGroupAction")
+    DataSourceDetector = apps.get_model("sentry", "DataSourceDetector")
+    DetectorWorkflow = apps.get_model("sentry", "DetectorWorkflow")
+    WorkflowDataConditionGroup = apps.get_model("sentry", "WorkflowDataConditionGroup")
+
+    for rule in RangeQuerySetWrapperWithProgressBarApprox(AlertRule.objects.all()):
```

Comment:
if so, i wonder how we could decompose the loop body into a method that's reusable for the migration and the API 🤔

Reply:
We don't typically share code between migrations and real code